Add controllers

churmantey opened the pull request:
ТЗ 14 спринта, на ревью.
Начал было делать сервисы и хранилища для ItemRequest и Booking, и только потом увидел, что их делать в этом спринте не надо. Но заготовки убирать не стал - думаю, пригодятся :)

churmantey committed:
in progress

in progress

Initial commit

churmantey committed:
Initial commit

churmantey committed:
in progress

churmantey committed:
in progress

churmantey committed:
userStorage, поиск пользователя по email

churmantey committed:
изменена структура пакетов, + itemRepository

churmantey committed:
Финальное ТЗ 14 спринта

churmantey committed:
Финальное ТЗ 14 спринта

churmantey committed:
Финальное ТЗ 14 спринта, корректировка

TarasovSam commented on pom.xml:
```diff
@@ -56,10 +56,28 @@
 			<artifactId>spring-boot-starter-validation</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>org.mapstruct</groupId>
```
MapStruct - шикарно!)

TarasovSam commented on pom.xml:
```diff
@@ -56,10 +56,28 @@
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct</artifactId>
+			<version>1.6.0.Beta2</version>
```
Версии лучше перенести в блок properties, так помнит станет более читаемым и поддерживаемым (все версии в одном месте)

churmantey replied:
Точно! перенес.

TarasovSam commented on src/main/java/ru/practicum/shareit/general/BaseRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.List;
+
+public interface BaseRepository<T> {
```
Хорошая идея, так же как и с IdGenerator :)

TarasovSam commented on src/main/java/ru/practicum/shareit/item/Item.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Positive
+    private Long owner; // id владельца
+    private Long request; // id запроса
```
Раздели поля пустой строкой, чтобы общий стиль класса был

churmantey replied:
Исправил

TarasovSam commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,65 @@
+    private final ItemService itemService;
+
+    @PostMapping
+    public ItemDto createItem(@RequestHeader("X-Sharer-User-Id") Long userId,
```
"X-Sharer-User-Id" вынеси в константу и используй по всему приложению

churmantey replied:
Хорошая идея, спасибо! 
Применил.

TarasovSam commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,65 @@
+                              @PathVariable Long itemId,
+                              @Valid @RequestBody ItemDto itemDto) {
+        log.info("UPDATE item {}, user {}, item data {}", itemId, userId, itemDto);
+        ItemDto existingItem = itemService.getItem(itemId);
```
Получение объекта, проверка - это бизнес логика, ее лучше убрать из контроллера в метод сервиса

churmantey replied:
Согласен, не додумал... 
И, кстати, после исправления получилось меньше обращений к хранилищу.

TarasovSam commented on src/main/java/ru/practicum/shareit/item/repository/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,72 @@
+        if (itemId == null) {
+            throw new ValidationException("Item id of null received");
+        }
+        if (itemStorage.containsKey(itemId)) return itemStorage.get(itemId);
```
Эту проверку можно прописать несколько по-другому используя orElseThrow()
Аналогично и в других так же местах

churmantey replied:
Спасибо! Применил!

TarasovSam commented on src/main/java/ru/practicum/shareit/user/User.java:
```diff
@@ -1,7 +1,21 @@
+
+    @Positive
+    private Long id;
+    private String name;
```
А имя не валидируется?

churmantey replied:
По тестам - нет. 
А по здравому смыслу - должно :) 
Поправил.

TarasovSam commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,15 @@
+    private String name;
+
+    @Email
+    @NotBlank
```
В dto точно нужна валидация?

churmantey replied:
Вот тут я пока не сложил для себя однозначной картины...
По ТЗ email у пользователя должен быть. 
При его создании я применяю для dto аннотацию @Valid:
отсечь ошибки на самой ранней стадии, на веб-слое, мне показалось правильным. 
С другой стороны, для обновления пользователя используется тот же dtо, и там уже поля email может не быть. 
Тут хочу у тебя совета попросить, как правильнее?

TarasovSam replied:
Тут несколько способов существует. Напишу основные два
1. в dto не валидировать через аннотацию, а в коде. Тогда используется одна "универсальная" dto
2. разделить dto на dto и dtoUpdate. И у каждой валидацию прописать свою
А еще есть такая штука, как валидация по группам. Т.е. есть одна dto, но аннотации прописываются с параметрами (применять в зависимости от группы, грубо говоря)

TarasovSam commented on src/test/java/ru/practicum/shareit/ShareItTests.java:
```diff
@@ -6,8 +6,8 @@
 @SpringBootTest
 class ShareItTests {
```
Тестирование же в Postman проходит, верно?

churmantey replied:
Да, эта заготовка была в начальном репозитории. 
Я ее трогать не стал. Наверное, потом пригодится.

TarasovSam reviewed (commented):
Данила, привет! Проверил твое задание, оставил комментарии с тем, что можно улучшить и что нужно поправить)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку

PS классы-заготовки не проверял) Но измени в них что-то в следующем спринте, чтобы они в PR попали

churmantey committed:
Корректировки после ревью:
версии pom -> в раздел properties
все получения объектов -> в слое сервисов
оптимизиция получения объектов из хранилища
http заголовок для ID пользователя ->  в константу

TarasovSam commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -22,8 +21,10 @@ public class ItemController {
+
     @PostMapping
-    public ItemDto createItem(@RequestHeader("X-Sharer-User-Id") Long userId,
+    public ItemDto createItem(@RequestHeader(USER_ID_HEADER) Long userId,
```
Кстати, для эдпоинтов дополнительно можно прописать аннотацию @ResponseStatus, которая возвращает статус
Например, для метода создания @ResponseStatus(HttpStatus.CREATED)

TarasovSam commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -31,7 +31,6 @@ public UserDto createUser(@Valid @RequestBody UserDto userDto) {
-        UserDto existingUser = userService.getUser(userId);
         userDto.setId(userId);
         return userService.updateUser(userDto);
     }
```
Напишу тут, но относится к строке 25 и к твоему комментарию на счет совета
Вместо @Valid можно использовать такую аннотацию @validated(Create.class). Таким образом мы указываем, какую группу аннотаций следует учитывать. Create - это интерфейс, который ты создаешь самостоятельно, он будет пустым и служит лишь для указания группы) Аналогично следует поступить при обновлении, но там будет аннотация @validated(Update.class)
Для вещи тоже самое)
Попробуй, заработает или нет, т.к. бывают случаи, когда тесты не проходят. Если не пройдут напиши мне

TarasovSam reviewed (commented):
Данила, проверил твое задание, спасибо за работу и за то, что дорабатываешь код. Оставил еще комментарии, в том числе и к твоим комментариям, посмотри их)
Задание я принимаю, успеха в обучении и до следующего спринта)
